internal/provider/table_schema_resource.go

```diff
@@ -320,29 +344,29 @@ func toDateTimeFieldSpecs(fieldSpecs []dateTimeFieldSpec) []model.FieldSpec {
 			DataType:    fs.DataType,
 			Format:      fs.Format,
 			Granularity: fs.Granularity,
-			NotNull:     fs.NotNull.ValueBool(),
+			NotNull:     boolPtr(fs.NotNull.ValueBool()),
 		})
 	}
 	return pinotFieldSpecs
 }
 
 func setState(state *tableSchemaResourceModel, schema *model.Schema) {
 
-	dimensionFieldSpecs := make([]fieldSpec, len(schema.DimensionFieldSpecs))
+	dimensionFieldSpecs := make([]dimensionFieldSpec, len(schema.DimensionFieldSpecs))
 	for i, fs := range schema.DimensionFieldSpecs {
-		dimensionFieldSpecs[i] = fieldSpec{
+		dimensionFieldSpecs[i] = dimensionFieldSpec{
 			Name:     fs.Name,
 			DataType: fs.DataType,
-			NotNull:  basetypes.NewBoolValue(fs.NotNull),
+			NotNull:  basetypes.NewBoolValue(*fs.NotNull),
 		}
 	}
```

Comment:
if we use the ValueBoolPointer type from base types as mentioned in the other comment, we can define this here as:
```go
func setState(state *tableSchemaResourceModel, schema *model.Schema) {

	dimensionFieldSpecs := make([]dimensionFieldSpec, len(schema.DimensionFieldSpecs))
	for i, fs := range schema.DimensionFieldSpecs {
		dimensionFieldSpecs[i] = dimensionFieldSpec{
			Name:     fs.Name,
			DataType: fs.DataType,
			NotNull:  basetypes.NewBoolPointerValue(fs.NotNull),
		}
	}
```